### DISCUSSION TITLE
Direct to display rendering

### DISCUSSION THREAD

**Discussion by brunorzn** [vsg-dev/VulkanSceneGraph#875] [Q&A]
Hello,

I would like to render my VSG Application without a X Server, for an embedded system (Raspberry PI).

I think it should be possible using VK_KHR_display ? But I didn't find the correct way to make this work. 

Any tip / idea ? 

Thank you.

**Comment by robertosfield** [comment]
I am not familiar with the extension.  Have you come across examples of how to implement support?

**Reply by brunorzn** [reply]
Not yet, I'm trying to have one Vulkan example working but I'm still looking for it.

This repo pops often: https://github.com/nvpro-samples/gl_render_vk_ddisplay

But this [thread](https://forums.raspberrypi.com/viewtopic.php?t=288044) has been the most promising so far. 

vulkaninfo works in the console, without X, so we should be able to set VK_KHR_display and set a proper physical device ? It's just a guess.

I will continue investigating.

**Reply by robertosfield** [reply]
I don't know if it helps but the vsgheadless example doesn't need any Windowing, and just rendering to a vkFramebuffer and copies the image on each frame to write it out to disk.

Others in the community have Raspberry Pi so perhaps they make have some insights, though feedback I've got is with xcb server side.  Do you have a X server on your machine to test against?

**Reply by brunorzn** [reply]
Yes, all examples and my vsg app work on the X server.

Yes I've been looking to vsgheadless, I think it's the closest thing we have in the examples.

The Vulkan Examples of Sacha Willems also work. There is a CMake option: D_DIRECT2DISPLAY.

I'm trying to run the Triangle example which segfaults when this option is set (and works great on a X server when it's not).

It for intsance does this:

#elif defined(_DIRECT2DISPLAY)
	instanceExtensions.push_back(VK_KHR_DISPLAY_EXTENSION_NAME);

**Comment by robertosfield** [comment]
For others looking in on this thread, I've gone a few quick checks on the VK_KHR_display extension.  First up I've run:

~~~ sh
 vsgdeviceselection -e 
~~~

On my system and amount the vsg::enumerateInstanceExtensionProperties() list I see:

    extensionName = VK_KHR_display, specVersion = 23

Right alongside the xcb and x11 extensions so even on my desktop system with AMD 5700G + Mesa drivers it looks like it may be supported :-)

The Khronos page on VK_KHR_display is:

https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_display.html

This provides a list of extensions for creating the vkSurface and the equivalent to the window, so perhaps a subclass from vsg::Surface and vsg::Window would be possible using these extensions in a similar way to the Xcb support is done, but simpler as it'd be missing all the event handling code and talking to the xcb server.
